### PR TITLE
JSDK-2911 Update MediaDevices example to use track.restart

### PR DESCRIPTION
**Pull request by PikaJoyce**
[JIRA Ticket](https://issues.corp.twilio.com/browse/JSDK-2911)

This PR is for updating the Media Devices example to use the latest SDK track.restart method. 

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: examples/mediadevices/src/helpers.js [hunk 1]
```diff
@@ -30,60 +30,24 @@ function applyAudioOutputDeviceSelection(deviceId, audio) {
...
-  }).then(function(localTrack) {
-    localTrack.attach(audio);
-    if (room) {
-      return room.localParticipant.publishTrack(localTrack);
-    }
-  }).catch(function(error) {
-    console.log('applyAudioInputDeviceSelection failed:', error);
+function applyAudioInputDeviceSelection(deviceId, audioTrack) {
+  return audioTrack.restart({ deviceId: deviceId }).catch(function(error) {
+    'LocalAudioTrack restart failed: ', error
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/helpers.js:38]:
```suggestion
    console.log('applyAudioInputDeviceSelection failed:', error);
```

File: examples/mediadevices/src/helpers.js [hunk 2]
```diff
@@ -30,60 +30,24 @@ function applyAudioOutputDeviceSelection(deviceId, audio) {
...
-  }).then(function(localTrack) {
-    localTrack.attach(video);
-    if (room) {
-      return room.localParticipant.publishTrack(localTrack);
-    }
-  }).catch(function(error) {
-    console.log('applyVideoInputDeviceSelection failed:', error);
+function applyVideoInputDeviceSelection(deviceId, videoTrack) {
+  return videoTrack.restart({ deviceId: deviceId }).catch(function(error) {
+    'LocalVideoTrack restart failed: ', error
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/helpers.js:50]:
```suggestion
    console.log('applyVideoInputDeviceSelection failed:', error);
```

File: examples/mediadevices/src/index.js [hunk 3]
```diff
@@ -119,20 +119,19 @@ function participantDisconnected(participant) {
...
   var waveformContainer = document.querySelector('div#audioinputwaveform');
   if (event) {
     event.preventDefault();
     event.stopPropagation();
   }
-
-  await applyAudioInputDeviceSelection(deviceSelections.audioinput.value, audio, someRoom);
-
-  if (audio.srcObject) {
+  const [localAudioTrackPublication] = Array.from(someRoom.localParticipant.audioTracks.values());
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/index.js:129]: This will break device selection when you have not yet connected to a Room.

File: examples/mediadevices/src/index.js [hunk 4]
```diff
@@ -141,9 +140,14 @@ async function applyAudioInputDeviceChange(event) {
 
 // reads selected video input, and updates preview and room to use the device.
 function applyVideoInputDeviceChange(event) {
+  const [localVideoTrackPublication] = Array.from(someRoom.localParticipant.videoTracks.values());
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/index.js:143]: This will break device selection when you have not yet connected to a Room.

File: examples/mediadevices/src/helpers.js [hunk 5]
```diff
@@ -30,60 +30,24 @@ function applyAudioOutputDeviceSelection(deviceId, audio) {
...
-  }).then(function(localTrack) {
-    localTrack.attach(audio);
-    if (room) {
-      return room.localParticipant.publishTrack(localTrack);
-    }
-  }).catch(function(error) {
-    console.log('applyAudioInputDeviceSelection failed:', error);
+function applyAudioInputDeviceSelection(deviceId, audioTrack) {
+  return audioTrack.restart({ deviceId: deviceId }).catch(function(error) {
+    console.log('LocalAudioTrack restart failed: ', error);
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/helpers.js:38]:
```suggestion
    console.log('applyAudioInputDeviceSelection failed:', error);
```

File: examples/mediadevices/src/helpers.js [hunk 6]
```diff
@@ -30,60 +30,24 @@ function applyAudioOutputDeviceSelection(deviceId, audio) {
...
-  }).then(function(localTrack) {
-    localTrack.attach(video);
-    if (room) {
-      return room.localParticipant.publishTrack(localTrack);
-    }
-  }).catch(function(error) {
-    console.log('applyVideoInputDeviceSelection failed:', error);
+function applyVideoInputDeviceSelection(deviceId, videoTrack) {
+  return videoTrack.restart({ deviceId: deviceId }).catch(function(error) {
+    console.log('LocalVideoTrack restart failed: ', error);
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/helpers.js:50]:
```suggestion
    console.log('applyVideoInputDeviceSelection failed:', error);
```

File: examples/mediadevices/src/index.js [hunk 7]
```diff
@@ -119,20 +119,22 @@ function participantDisconnected(participant) {
...
-
-  await applyAudioInputDeviceSelection(deviceSelections.audioinput.value, audio, someRoom);
-
-  if (audio.srcObject) {
-    var canvas = waveformContainer.querySelector('canvas');
-    waveform.setStream(audio.srcObject);
+  if(someRoom) {
+    [localAudioTrackPublication] = Array.from(someRoom.localParticipant.audioTracks.values());
+  }
+  if (localAudioTrackPublication) {
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/index.js:133]: This still doesn't work if you want to change and apply devices before joining a Room.

File: examples/mediadevices/src/index.js [hunk 8]
```diff
@@ -141,9 +143,21 @@ async function applyAudioInputDeviceChange(event) {
...
+    event.stopPropagation();
+  }
+  if(someRoom) {
+    [localVideoTrackPublication] = Array.from(someRoom.localParticipant.videoTracks.values());
+  }
   try {
-    var video = document.querySelector('video#videoinputpreview');
-    applyVideoInputDeviceSelection(deviceSelections.videoinput.value, video, someRoom);
+    const video = document.querySelector('video#videoinputpreview');
+    applyVideoInputDeviceSelection(deviceSelections.videoinput.value, localVideoTrackPublication.track)
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/index.js:156]: This still doesn't work if you want to change and apply devices before joining a Room. Also, `localVideoTrackPublication.track` will raise an error if you haven't joined the Room (publication will be null).

**Comment by PikaJoyce** [comment]
@manjeshbhargav this is how it looks like now! Works great on FF.
mediaDevices

File: examples/mediadevices/src/helpers.js [hunk 9]
```diff
@@ -30,61 +30,25 @@ function applyAudioOutputDeviceSelection(deviceId, audio) {
...
+ * @param {string} kind - kind of Track
  * @returns {Promise<void>}
  */
-function applyAudioInputDeviceSelection(deviceId, audio, room) {
-  // Stop and unpublish the existing LocalAudioTrack.
-  if (room) {
-    room.localParticipant.audioTracks.forEach(function(publication) {
-      publication.track.stop();
-      publication.unpublish();
+function applyInputDeviceSelection(deviceId, localTrack, kind) {
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/helpers.js:37]: I think you can simplify this function to:
```js
function applyInputDeviceSelection(deviceId, localTrack, kind) {
  var constraints = { deviceId: { exact: deviceId } };
  if (localTrack) {
    return localTrack.restart(constraints).then(function() {
      return localTrack;
    });
  }
  return kind === 'audio'
    ? Video.createLocalAudioTrack(constraints); 
    : Video.createLocalVideoTrack(constraints);
}
```

File: examples/mediadevices/src/index.js [hunk 10]
```diff
@@ -119,31 +122,42 @@ function participantDisconnected(participant) {
   participantDiv.parentNode.removeChild(participantDiv);
 }
 
-// reads selected audio input, and updates preview and room to use the device.
+// // reads selected audio input, and updates preview and room to use the device.
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/index.js:125]:
```suggestion
// reads selected audio input, and updates preview and room to use the device.
```

File: examples/mediadevices/src/index.js [hunk 11]
```diff
@@ -1,19 +1,22 @@
...
 var helpers = require('./helpers');
 var waveform = require('../../util/waveform');
-var applyAudioInputDeviceSelection = helpers.applyAudioInputDeviceSelection;
+var applyInputDeviceSelection = helpers.applyInputDeviceSelection;
 var applyAudioOutputDeviceSelection = helpers.applyAudioOutputDeviceSelection;
-var applyVideoInputDeviceSelection = helpers.applyVideoInputDeviceSelection;
 const connectWithSelectedDevices = helpers.connectWithSelectedDevices;
 const connectOrDisconnect = document.querySelector('input#connectordisconnect');
 const getRoomCredentials = require('../../util/getroomcredentials');
+const Video = require('twilio/lib/rest/Video');
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/index.js:13]: Why is this required?

File: examples/mediadevices/src/index.js [hunk 12]
```diff
@@ -176,7 +193,7 @@ async function connectOrDisconnectRoom(event) {
   } else {
 
     const creds = await getRoomCredentials();
-    someRoom = await connectWithSelectedDevices(creds.token, deviceSelections.audioinput.value, deviceSelections.videoinput.value);
+    // someRoom = await Video.connect(creds.token, {tracks: [localVideoTrack, Video.LocalAudioTrack]});
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/index.js:196]:
```suggestion
    someRoom = await Video.connect(creds.token, {
      tracks: [localAudioTrack, localVideoTrack]
    });
```

File: examples/mediadevices/src/index.js [hunk 13]
```diff
@@ -119,31 +122,42 @@ function participantDisconnected(participant) {
...
     event.stopPropagation();
   }
-
-  await applyAudioInputDeviceSelection(deviceSelections.audioinput.value, audio, someRoom);
-
-  if (audio.srcObject) {
-    var canvas = waveformContainer.querySelector('canvas');
-    waveform.setStream(audio.srcObject);
+  try {
+    if(!localAudioTrack) {
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/index.js:133]:
```suggestion
```

File: examples/mediadevices/src/index.js [hunk 14]
```diff
@@ -119,31 +122,42 @@ function participantDisconnected(participant) {
...
+      localAudioTrack = await applyInputDeviceSelection(deviceSelections.audioinput.value, localAudioTrack, 'audio');
+      console.log('localAudioTrack after apply', localAudioTrack);
+    }
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/index.js:136]:
```suggestion
```

File: examples/mediadevices/src/index.js [hunk 15]
```diff
@@ -119,31 +122,42 @@ function participantDisconnected(participant) {
...
-
-  await applyAudioInputDeviceSelection(deviceSelections.audioinput.value, audio, someRoom);
-
-  if (audio.srcObject) {
-    var canvas = waveformContainer.querySelector('canvas');
-    waveform.setStream(audio.srcObject);
+  try {
+    if(!localAudioTrack) {
+      localAudioTrack = await applyInputDeviceSelection(deviceSelections.audioinput.value, localAudioTrack, 'audio');
+      console.log('localAudioTrack after apply', localAudioTrack);
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/index.js:135]:
```suggestion
```

File: examples/mediadevices/src/index.js [hunk 16]
```diff
@@ -119,31 +122,42 @@ function participantDisconnected(participant) {
...
+    }
+    await applyInputDeviceSelection(deviceSelections.audioinput.value, localAudioTrack, 'audio');
+    console.log('localAudioTrack exists',localAudioTrack);
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/index.js:138]:
```suggestion
```

File: examples/mediadevices/src/index.js [hunk 17]
```diff
@@ -119,31 +122,42 @@ function participantDisconnected(participant) {
...
+  if (event) {
+    event.preventDefault();
+    event.stopPropagation();
+  }
   try {
-    var video = document.querySelector('video#videoinputpreview');
-    applyVideoInputDeviceSelection(deviceSelections.videoinput.value, video, someRoom);
+    const video = document.querySelector('video#videoinputpreview');
+    localVideoTrack = await applyInputDeviceSelection(deviceSelections.videoinput.value, localVideoTrack, 'video').catch(err => (console.error(err)));
+    localVideoTrack.attach(video)
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/index.js:160]:
```suggestion
    localVideoTrack.attach(video);
```

File: examples/mediadevices/src/index.js [hunk 18]
```diff
@@ -119,31 +122,42 @@ function participantDisconnected(participant) {
...
+async function applyVideoInputDeviceChange(event) {
+  if (event) {
+    event.preventDefault();
+    event.stopPropagation();
+  }
   try {
-    var video = document.querySelector('video#videoinputpreview');
-    applyVideoInputDeviceSelection(deviceSelections.videoinput.value, video, someRoom);
+    const video = document.querySelector('video#videoinputpreview');
+    localVideoTrack = await applyInputDeviceSelection(deviceSelections.videoinput.value, localVideoTrack, 'video').catch(err => (console.error(err)));
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/index.js:159]:
```suggestion
    localVideoTrack = await applyInputDeviceSelection(deviceSelections.videoinput.value, localVideoTrack, 'video');
```

File: examples/mediadevices/src/index.js [hunk 19]
```diff
@@ -119,31 +122,42 @@ function participantDisconnected(participant) {
...
     event.preventDefault();
     event.stopPropagation();
   }
-
-  await applyAudioInputDeviceSelection(deviceSelections.audioinput.value, audio, someRoom);
-
-  if (audio.srcObject) {
-    var canvas = waveformContainer.querySelector('canvas');
-    waveform.setStream(audio.srcObject);
+  try {
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/index.js:132]:
```suggestion
```

File: examples/mediadevices/src/index.js [hunk 20]
```diff
@@ -119,31 +122,42 @@ function participantDisconnected(participant) {
...
+    await applyInputDeviceSelection(deviceSelections.audioinput.value, localAudioTrack, 'audio');
+    console.log('localAudioTrack exists',localAudioTrack);
+    const canvas = waveformContainer.querySelector('canvas');
+    waveform.setStream(new MediaStream([localAudioTrack.mediaStreamTrack]));
     if (!canvas) {
       waveformContainer.appendChild(waveform.element);
     }
+  } catch (error) {
+    console.log('audioInput apply failed:', error);
   }
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/index.js:146]:
```suggestion
```

File: examples/mediadevices/src/index.js (hunk 11 above)
Comment (PikaJoyce) [reply examples/mediadevices/src/index.js:13]: Oh, I believe that was added when I was trying to use Video.connect! WIll remove in the next commit. Thank you for catching that!

File: examples/mediadevices/src/index.js [hunk 21]
```diff
@@ -122,27 +120,18 @@ function participantDisconnected(participant) {
   participantDiv.parentNode.removeChild(participantDiv);
 }
 
-// // reads selected audio input, and updates preview and room to use the device.
+// Reads selected audio input, and updates preview and room to use the device.
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/index.js:123]:
```suggestion
// reads selected audio input, and updates preview and room to use the device.
```

File: examples/mediadevices/src/index.js [hunk 22]
```diff
@@ -156,15 +145,17 @@ async function applyVideoInputDeviceChange(event) {
   }
   try {
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/index.js:146]:
```suggestion
```

File: examples/mediadevices/src/index.js [hunk 23]
```diff
@@ -156,15 +145,17 @@ async function applyVideoInputDeviceChange(event) {
...
-    localVideoTrack.attach(video)
+    localVideoTrack = await applyInputDeviceSelection(deviceSelections.videoinput.value, localVideoTrack, 'video');
+    localVideoTrack.attach(video);
     if (event) {
       event.preventDefault();
       event.stopPropagation();
     }
   } catch (error) {
     console.log('videoInput apply failed:', error);
   }
```
Comment (manjeshbhargav) [inline examples/mediadevices/src/index.js:156]:
```suggestion
```
